ntcore/ds_client.py

```diff
@@ -1,21 +1,112 @@
-# validated: 0000-00-00 DS 000000000 cpp/DsClient.cpp cpp/DsClient.h
-'''----------------------------------------------------------------------------'''
-''' Copyright (c) FIRST 2017. All Rights Reserved.                             '''
-''' Open Source Software - may be modified and shared by FRC teams. The code   '''
-''' must be accompanied by the FIRST BSD license file in the root directory of '''
-''' the project.                                                               '''
-'''----------------------------------------------------------------------------'''
+# validated: 2017-12-09 DV 5ab20bb27c97 cpp/DsClient.cpp cpp/DsClient.h
+#----------------------------------------------------------------------------
+# Copyright (c) FIRST 2017. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+#----------------------------------------------------------------------------
+
+import json
+import threading
+
+from .support.compat import Condition
+from .support.safe_thread import SafeThread
+from .tcpsockets.tcp_connector import TcpConnector
 
 import logging
 logger = logging.getLogger('nt')
 
+
 class DsClient(object):
     
     def __init__(self, dispatcher, verbose=False):
-        pass
-    
+        self.m_dispatcher = dispatcher
+        self.verbose = verbose
+
+        self.m_active = False
+        self.m_owner = None  # type: SafeThread
+
+        self.m_mutex = threading.Lock()
+        self.m_cond = Condition(self.m_mutex)
+
+        self.m_port = None  # type: int
+        self.m_stream = None
+
     def start(self, port):
-        logging.warning("Dashboard client mode not currently implemented")
-    
+        with self.m_mutex:
+            self.m_port = port
+            if not self.m_active:
+                self.m_active = True
+                self.m_owner = SafeThread(target=self._thread_main, name='nt-dsclient')
+
     def stop(self):
-        pass
+        with self.m_mutex:
+            # Close the stream so the read (if any) terminates.
+            self.m_active = False
+            if self.m_stream:
+                self.m_stream.close()
+            self.m_cond.notify()
+
+    def _thread_main(self):
+        oldip = 0
+        connector = TcpConnector(verbose=False, timeout=1)
```

Comment:
I wonder if verbose should be set to self.verbose?

Reply:
Hm. Upstream completely silences this one altogether, so I was just following that.

This would get noisy if it can't connect to the DS.